File: app/src/main/java/it/bitprepared/prbm/mobile/activity/EntityActivity.java
/*   This file is part of PrbmMobile
 *
 *   PrbmMobile is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   PrbmMobile is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with PrbmMobile.  If not, see <http://www.gnu.org/licenses/>.
 */

package it.bitprepared.prbm.mobile.activity;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.google.gson.Gson;

import android.app.ActionBar;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.provider.OpenableColumns;
import android.util.Log;
import android.view.ContextThemeWrapper;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.TimePicker;
import android.widget.Toast;

import it.bitprepared.prbm.mobile.R;
import it.bitprepared.prbm.mobile.model.PrbmEntity;
import it.bitprepared.prbm.mobile.model.PrbmUnit;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;

/**
 * Class responsible for visualizing and modifying a single PrbmEntity
 */
public class EntityActivity extends Activity {

    /** Debug TAG */
    private final static String TAG = "EntityActivity";
    private static final int CAMERA_RESULT = 1005;
    private static final int GALLERY_RESULT = 1006;

    /** Linear layout left free to each Entity */
    private LinearLayout linFree = null;

    /** Reference to Entity Caption textbox */
    private EditText edtCaption = null;
    /** Reference to Entity Description textbox */
    private EditText edtDescription = null;
    /** Reference to Entity Time textbox */
    private TimePicker datTime = null;
    /** Reference to the ImageView */
    private ImageView imgCamera = null;

    /** Boolean flag for edit functionalities */
    private boolean edit = false;

    /** Image URI */
    private transient Uri capturedImageUri = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Intent intent = getIntent();
        edit = intent.getBooleanExtra("edit", false);

        // Setting up Home back button
        ActionBar bar = getActionBar();
        if (bar != null) {
            bar.setDisplayHomeAsUpEnabled(true);
        }

        // Inflating views
        setContentView(R.layout.activity_entity);
        linFree = findViewById(R.id.linearFreeEntity);
        ImageView imgBack = findViewById(R.id.imgEntity);
        TextView txtTitle = findViewById(R.id.txtEntityTitleAdd);
        datTime = findViewById(R.id.datTimeEntity);
        edtCaption = findViewById(R.id.edtCaption);
        edtDescription = findViewById(R.id.edtDescription);
        imgCamera = findViewById(R.id.imgCamera);

        PrbmEntity entity = UserData.getInstance().getEntity();
        if (entity != null) {
            imgBack.setImageResource(entity.getIdBackImage());
            entity.drawYourSelf(EntityActivity.this, linFree);
            txtTitle.setText(entity.getType());

            Gson gson = new Gson();
            Log.e("TAG", gson.toJson(entity));

            if (!edit) {
                // Setting current hour
                Calendar c = Calendar.getInstance(getResources().getConfiguration().locale);
                datTime.setCurrentHour(c.get(Calendar.HOUR_OF_DAY));
                datTime.setCurrentMinute(c.get(Calendar.MINUTE));
                datTime.setIs24HourView(true);
            } else {
                edtCaption.setText(entity.getCaption());
                edtDescription.setText(entity.getDescription());
                if (!entity.getPictureName().isEmpty()) {
                    capturedImageUri = entity.getPictureURI();
                    imgCamera.setVisibility(View.VISIBLE);

                    Glide
                            .with(this)
                            .load(capturedImageUri)
                            .apply(new RequestOptions().override(600, 300).centerCrop())
                            .into(imgCamera);
                }

                // Restoring timestamp to TimePicker
                SimpleDateFormat dateFormat = new SimpleDateFormat(
                    "yyyy-MM-dd HH:mm:ss", getResources().getConfiguration().locale);
                try {
                    Date date = dateFormat.parse(entity.getTimestamp());
                    Calendar cal = Calendar.getInstance(getResources().getConfiguration().locale);
                    cal.setTime(date);
                    datTime.setCurrentHour(cal.get(Calendar.HOUR_OF_DAY));
                    datTime.setCurrentMinute(cal.get(Calendar.MINUTE));
                } catch (ParseException e) {
                }
                entity.restoreFields(this, linFree);
            }
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.prbm_entity, menu);
        return true;
    }


    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        int id = item.getItemId();
        if (id == R.id.discard || id == android.R.id.home) {

            AlertDialog.Builder build = new AlertDialog.Builder(EntityActivity.this);
            build.setTitle(R.string.confirmation);
            build.setMessage(R.string.are_you_sure);
            build.setPositiveButton(R.string.ok, (dialog, which) -> {
                PrbmEntity entity = UserData.getInstance().getEntity();
                entity.setPictureName("");
                finish();
            });
            build.setNegativeButton(R.string.abort, (dialog, which) -> {
            });
            build.show();
            return true;
        } else if (id == R.id.save) {

            // Checking if caption is empty
            if (edtCaption.getText().length() == 0) {
                AlertDialog.Builder alert = new AlertDialog.Builder(
                    new ContextThemeWrapper(this, R.style.AlertDialogCustom));
                alert.setTitle(R.string.fields_incomplete);
                alert.setMessage(getString(R.string.you_must_insert_caption));
                alert.setIcon(R.drawable.ic_alert_black_48dp);
                alert.setPositiveButton(R.string.ok,
                        (dialog, whichButton) -> dialog.dismiss());
                alert.show();
            } else {
                PrbmEntity entity = UserData.getInstance().getEntity();
                if (entity != null) {
                    entity.saveFields(EntityActivity.this, linFree);
                    entity.setCaption(edtCaption.getText().toString());
                    entity.setDescription(edtDescription.getText().toString());
                    Calendar c = Calendar.getInstance(getResources().getConfiguration().locale);

                    // Saving timestamp. Date set at Unix epoch
                    c.set(1970, 1, 1, datTime.getCurrentHour(), datTime.getCurrentMinute());
                    SimpleDateFormat dateFormat = new SimpleDateFormat(
                        "yyyy-MM-dd HH:mm:ss", getResources().getConfiguration().locale);
                    entity.setTimestamp(dateFormat.format(c.getTime()));

                    if (!edit) {
                        PrbmUnit involved = UserData.getInstance().getUnit();
                        involved.addEntity(entity, UserData.getInstance().getColumn());
                    }
                    setResult(RESULT_OK);
                    if (PrbmAddEntityActivity.self != null) {
                        PrbmAddEntityActivity.self.finish();
                    }
                    finish();
                }
            }
            return true;
        } else if (id == R.id.pic) {
            PrbmEntity entity = UserData.getInstance().getEntity();
            CharSequence[] itemadd = {"Scatta una foto",
                                      "Scegli dalla Galleria"};
            CharSequence[] itemadddelete = {"Scatta una foto",
                                            "Scegli dalla Galleria", "Rimuovi foto"};
            boolean deleteflag = (entity != null && !entity.getPictureName().isEmpty());
            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setIcon(R.drawable.ic_camera_iris_black_36dp);
            builder.setTitle("Fotografia");
            builder.setItems((deleteflag ? itemadddelete : itemadd),
                             (dialog, which) -> {
                                 switch (which) {
                                     case 0:
                                         takePicture();
                                         dialog.dismiss();
                                         break;
                                     case 1:
                                         chooseFromGallery();
                                         dialog.dismiss();
                                         break;
                                     case 2:
                                         entity.setPictureName("");
                                         imgCamera.setVisibility(View.GONE);
                                         dialog.dismiss();
                                         break;
                                     default:
                                         break;
                                 }
                             });
            AlertDialog alert = builder.create();
            alert.setCancelable(true);
            alert.show();
        }
        return super.onOptionsItemSelected(item);
    }

    private String getFilenameFromURI(Uri uri) {
        if (uri == null) {
            return "";
        }
        String result = null;
        if (uri.getScheme().equals("content")) {
            Cursor cursor = getContentResolver().query(uri, null, null, null, null);
            try {
                if (cursor != null && cursor.moveToFirst()) {
                    result = cursor.getString(cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME));
                }
            } finally {
                cursor.close();
            }
        }
        if (result == null) {
            result = uri.getPath();
            int cut = result.lastIndexOf('/');
            if (cut != -1) {
                result = result.substring(cut + 1);
            }
        }
        return result;
    }

    private void chooseFromGallery() {
        Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
        intent.setType("image/*");
        startActivityForResult(intent, GALLERY_RESULT);
    }

    private void takePicture() {

        Calendar cal = Calendar.getInstance();
        File root = android.os.Environment.getExternalStorageDirectory();
        File dir = new File(root.getAbsolutePath() + "/PRBM");
        dir.mkdirs();
        String picname = cal.getTimeInMillis() + ".jpg";
        File destfile = new File(dir, (cal.getTimeInMillis() + ".jpg"));

        if (!destfile.exists()) {
            try {
                destfile.createNewFile();
            } catch (IOException e) {
                e.printStackTrace();
            }
        } else {
            destfile.delete();
            try {
                destfile.createNewFile();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        capturedImageUri = Uri.fromFile(destfile);
        Intent i = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        i.putExtra(MediaStore.EXTRA_OUTPUT, capturedImageUri);
        startActivityForResult(i, CAMERA_RESULT);
    }

    public void copy(File src, File dst) throws IOException {
        InputStream in = new FileInputStream(src);
        OutputStream out = new FileOutputStream(dst);
        // Transfer bytes from in to out
        byte[] buf = new byte[1024];
        int len;
        while ((len = in.read(buf)) > 0) {
            out.write(buf, 0, len);
        }
        in.close();
        out.close();
    }

    public static String getRealPathFromUri(Context context, Uri contentUri) {
        Cursor cursor = null;
        try {
            String[] proj = {MediaStore.Images.Media.DATA};
            cursor = context.getContentResolver().query(contentUri, proj, null, null, null);
            int column_index = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA);
            cursor.moveToFirst();
            return cursor.getString(column_index);
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }
    }

    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        PrbmEntity entity = UserData.getInstance().getEntity();

        if (requestCode == CAMERA_RESULT) {
            imgCamera.setVisibility(View.VISIBLE);
            Glide
                    .with(this)
                    .load(capturedImageUri)
                    .apply(new RequestOptions().override(600, 300).centerCrop())
                    .into(imgCamera);

            entity.setPictureName(getFilenameFromURI(capturedImageUri));
            Log.d(TAG, "capturedImage " + capturedImageUri.getPath());
        } else if (requestCode == GALLERY_RESULT) {
            if (resultCode != RESULT_OK) {
                return;
            }
            if (data == null) {
                Toast.makeText(this, "Impossibile caricare l'immagine", Toast.LENGTH_SHORT).show();
                return;
            }
            capturedImageUri = data.getData();
            imgCamera.setVisibility(View.VISIBLE);
            Glide
                    .with(this)
                    .load(data.getData())
                    .apply(new RequestOptions().override(600, 300).centerCrop())
                    .into(imgCamera);

            Calendar cal = Calendar.getInstance();
            File root = android.os.Environment.getExternalStorageDirectory();
            File dir = new File(root.getAbsolutePath() + "/PRBM");
            dir.mkdirs();
            String picname = cal.getTimeInMillis() + ".jpg";
            File destfile = new File(dir, (cal.getTimeInMillis() + ".jpg"));

            if (!destfile.exists()) {
                try {
                    destfile.createNewFile();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            } else {
                destfile.delete();
                try {
                    destfile.createNewFile();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            try {
                copy(new File(getRealPathFromUri(this, capturedImageUri)), destfile);
            } catch (IOException e) {
                e.printStackTrace();
            }
            entity.setPictureName(picname);
        }
    }
}

File: app/src/main/java/it/bitprepared/prbm/mobile/activity/EntityActivity.kt
package it.bitprepared.prbm.mobile.activity

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.text.Editable
import android.text.InputType
import android.text.TextWatcher
import android.util.Log
import android.util.TypedValue
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.PickVisualMediaRequest
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.cardview.widget.CardView
import androidx.core.content.FileProvider
import androidx.core.view.get
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.textfield.TextInputLayout
import com.google.android.material.timepicker.MaterialTimePicker
import com.google.android.material.timepicker.TimeFormat
import it.bitprepared.prbm.mobile.R
import it.bitprepared.prbm.mobile.databinding.ActivityEntityBinding
import it.bitprepared.prbm.mobile.databinding.EditGenericFieldBinding
import it.bitprepared.prbm.mobile.model.PrbmEntityField
import kotlinx.coroutines.launch
import java.io.File
import java.io.IOException
import java.time.LocalTime


/**
 * Class responsible for visualizing and modifying a single PrbmEntity
 */
class EntityActivity : AppCompatActivity() {
  private lateinit var binding: ActivityEntityBinding
  private val viewModel: EntityViewModel by viewModels()

  private lateinit var latestUri: Uri
  private lateinit var pickMedia: ActivityResultLauncher<PickVisualMediaRequest>
  private lateinit var takePicture: ActivityResultLauncher<Uri>

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    binding = ActivityEntityBinding.inflate(layoutInflater)
    setContentView(binding.root)

    lifecycleScope.launch {
      repeatOnLifecycle(Lifecycle.State.STARTED) {
        viewModel.modelState.collect { state ->
          if (state.saveReady) {
            finish()
          } else {
            binding.topAppBar.title = state.typeDescription
            binding.editTime.text = state.time
            binding.edtTitle.setTextIfDifferent(state.title)
            binding.edtDescription.setTextIfDifferent(state.description)
            renderFields(state.fields, state.fieldValues)
            binding.topAppBar.menu[0].isVisible = state.isEditing
            binding.topAppBar.setOnMenuItemClickListener { menuItem ->
              when (menuItem.itemId) {
                R.id.delete -> {
                  confirmDelete()
                  true
                }
                else -> false
              }
            }
            binding.linGallery.removeAllViews()
            state.imageUris.forEach { image ->
              val px8 = TypedValue.applyDimension(
                TypedValue.COMPLEX_UNIT_DIP, 8F, this@EntityActivity.getResources().displayMetrics
              ).toInt()
              val px16 = TypedValue.applyDimension(
                TypedValue.COMPLEX_UNIT_DIP, 16F, this@EntityActivity.getResources().displayMetrics
              ).toInt()
              val px80 = TypedValue.applyDimension(
                TypedValue.COMPLEX_UNIT_DIP, 80F, this@EntityActivity.getResources().displayMetrics
              ).toInt()
              val cardView = CardView(this@EntityActivity).apply {
                radius = px8.toFloat()
                layoutParams = LinearLayout.LayoutParams(px80, px80).apply {
                  marginStart = px16
                }
              }
              val imageView = ImageView(this@EntityActivity).apply {
                layoutParams = ViewGroup.LayoutParams(
                  ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT
                )
              }
              cardView.setOnClickListener {
                MaterialAlertDialogBuilder(this@EntityActivity)
                  .setTitle("Rimuovere l'immagine?")
                  .setIcon(R.drawable.ic_delete)
                  .setMessage("Sei sicuro di voler cancellare questa immagine dall'osservazione?")
                  .setNegativeButton(R.string.abort) { dialog, _ -> dialog.dismiss() }
                  .setPositiveButton(R.string.delete) { _, _ -> viewModel.removeImage(image) }
                  .create()
                  .show()
              }
              cardView.addView(imageView)
              Glide.with(this@EntityActivity).load(image)
                .apply(RequestOptions().override(80, 80).centerCrop()).into(imageView)
              binding.linGallery.addView(cardView)
            }
          }
        }
      }
    }
    binding.edtTitle.addTextChangedListener(object : TextWatcher {
      override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
      override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
      override fun afterTextChanged(s: Editable?) {
        viewModel.updateTitle(s.toString())
      }
    })
    binding.edtDescription.addTextChangedListener(object : TextWatcher {
      override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
      override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
      override fun afterTextChanged(s: Editable?) {
        viewModel.updateDescription(s.toString())
      }
    })
    binding.editTime.setOnClickListener {
      val time = LocalTime.parse(binding.editTime.getText().toString())
      val timePicker =
        MaterialTimePicker.Builder().setTimeFormat(TimeFormat.CLOCK_24H).setHour(time.hour)
          .setMinute(time.minute).setTitleText(R.string.pick_unit_time).build()
      timePicker.addOnPositiveButtonClickListener { _ ->
        viewModel.updateTime(timePicker.hour, timePicker.minute)
      }
      timePicker.show(supportFragmentManager, "timePicker")
    }
    binding.topAppBar.setNavigationOnClickListener {
      confirmExit()
    }
    binding.btnSaveEntity.setOnClickListener {
      viewModel.saveEntity()
    }
    binding.cardCamera.setOnClickListener {
      takePicture()
    }
    binding.cardGallery.setOnClickListener {
      selectFromGallery()
    }

    pickMedia = registerForActivityResult(ActivityResultContracts.PickVisualMedia()) { uri ->
      if (uri != null) {
        viewModel.addImage(uri.toString(), getFilenameFromUri(uri))
      }
    }
    takePicture = registerForActivityResult(ActivityResultContracts.TakePicture()) { result ->
      if (result) {
        viewModel.addImage(latestUri.toString(), getFilenameFromUri(latestUri))
      }
    }
  }

  private fun renderFields(
    fields: List<PrbmEntityField>,
    fieldValues: Map<String, String>
  ) {
    if (binding.linFree.childCount == 0) {
      fields.forEach { field ->
        val fieldBinding = EditGenericFieldBinding.inflate(layoutInflater, binding.linFree, false)
        fieldBinding.textinputGeneric.hint = field.name
        fieldBinding.textinputGeneric.helperText = field.hint
        fieldBinding.edtFieldGeneric.inputType = if (field.type == "number") {
          // TODO This doesn't work, fix it
          InputType.TYPE_CLASS_NUMBER or InputType.TYPE_CLASS_TEXT
        } else {
          InputType.TYPE_CLASS_TEXT
        }
        binding.linFree.addView(fieldBinding.root)
        fieldBinding.edtFieldGeneric.setText(fieldValues.getOrDefault(field.name, ""))
        fieldBinding.edtFieldGeneric.addTextChangedListener(object : TextWatcher {
          override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
          override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
          override fun afterTextChanged(s: Editable?) {
            viewModel.updateFieldValue(field.name, s.toString())
          }
        })
      }
    } else {
      fields.forEachIndexed { index, field ->
        val editText = (binding.linFree.getChildAt(index) as TextInputLayout).editText
        editText?.setTextIfDifferent(fieldValues.getOrDefault(field.name, ""))
      }
    }
  }

  private fun confirmExit() =
    MaterialAlertDialogBuilder(this).setTitle(R.string.confirmation).setIcon(R.drawable.ic_alert)
      .setMessage(R.string.are_you_sure)
      .setPositiveButton(getString(R.string.exit_without_save)) { _, _ ->
        finish()
      }.setNegativeButton(R.string.abort) { _, _ -> }.create().show()

  private fun confirmDelete() =
    MaterialAlertDialogBuilder(this).setTitle(R.string.confirm_delete).setIcon(R.drawable.ic_delete)
      .setMessage(R.string.are_you_sure_delete_entity)
      .setPositiveButton(getString(R.string.delete)) { _, _ ->
        viewModel.deleteEntity()
      }.setNegativeButton(R.string.abort) { _, _ -> }.create().show()

  private fun getNewImageFilename() = "prbm_${System.currentTimeMillis()}.jpg"

  private fun getNewImageUri(): Uri =
    if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
      getNewImageUriWithFileProvider()
    } else {
      getNewImageUriWithFileUri()
    }

  private fun getNewImageUriWithFileProvider(): Uri {
    val tempImagesDir = File(
      applicationContext.filesDir, //this function gets the external cache dir
      "prbmm_images"
    ) //gets the directory for the temporary images dir
    tempImagesDir.mkdir() //Create the temp_images dir
    val tempImage = File(
      tempImagesDir, //prefix the new abstract path with the temporary images dir path
      getNewImageFilename()
    ) //gets the abstract temp_image file name

    return FileProvider.getUriForFile(
      applicationContext,
      "it.bitprepared.prbm.mobile.provider",
      tempImage
    )
  }

  private fun getNewImageUriWithFileUri(): Uri {
    val fileName = getNewImageFilename()
    val root = Environment.getExternalStorageDirectory()
    val imagesFolder = File(root, "prbm_images").apply {
      mkdirs()
    }
    return Uri.fromFile(File(imagesFolder, fileName))
  }

  private fun takePicture() {
    latestUri = getNewImageUri()
    takePicture.launch(latestUri)
  }

  private fun selectFromGallery() {
    pickMedia.launch(PickVisualMediaRequest(ActivityResultContracts.PickVisualMedia.ImageOnly))
  }

  private fun getFilenameFromUri(uri: Uri): String? =
    if (uri.lastPathSegment != null && uri.lastPathSegment!!.endsWith(".jpg")) {
      uri.lastPathSegment
    } else {
      getNewImageFilename()
    }
}
